napari/_vispy/volume.py

```diff
@@ -58,6 +58,23 @@
 
 frag_dict['minip'] = MINIP_FRAG_SHADER
 
+AVG_SNIPPETS = dict(
+    before_loop="""
+        float sum = 0.0; // The sum of encountered values
+        """,
+    in_loop="""
+        sum += val;
+        """,
```

Comment:
I wonder if he we want to use an iterative moving average to prevent any roll over effects? I don't have experience here, but wonder if @jni or @tlambert03 have recommendation?

Reply:
Formula here http://datagenetics.com/blog/november22017/index.html#:~:text=It's%20often%20simply%20called%20the,Done!

Reply:
`mean_n = mean_(n-1) + (x_n - mean_(n-1)) / n` this will have greater numerical stability than just summing all the `x_n` first and then dividing by `N`

Reply:
no particular insight here ... would these only be different in cases of overflow?

Reply:
>  would these only be different in cases of overflow?

yes, they are "mathematically" identical.

Reply:
or at least should be if i wrote down the right equation ;-)

Reply:
This definitely looks important and worth getting right before merging - I came across the same thing for the variance but hadn't thought about it for the average. Will implement and test this tomorrow sometime

Reply:
> > would these only be different in cases of overflow?
> 
> yes, they are "mathematically" identical.

Overflow is the wrong word here, not sure whether @tlambert03 meant it as "exceeding the maximum value representable by a dtype". This is an issue of precision, not overflow. Roughly speaking, you don't want to be adding floating point numbers together that are on vastly different ranges, as is the case for a total sum vs an individual element. Here's a very good explanation for this:

https://blog.demofox.org/2017/11/21/floating-point-precision/

and for the pure maths types:

http://webhome.auburn.edu/~tamtiny/lecture%207.pdf

;)

Reply:
okay, updated and everything seems to look as expected with the incremental mean!

![avg_intensity_projection_phages](https://user-images.githubusercontent.com/7307488/99438928-cd934e80-2914-11eb-8240-c2f1fdaba3ed.gif)